src/Bridge/Symfony/Bundle/DataProvider/TraceableChainCollectionDataProvider.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Symfony\Bundle\DataProvider;
+
+use ApiPlatform\Core\DataProvider\ChainCollectionDataProvider;
+use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\ContextAwareCollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use ApiPlatform\Core\Exception\ResourceClassNotSupportedException;
+
+/**
+ * @author Anthony GRASSIOT <antograssiot@free.fr>
+ */
+final class TraceableChainCollectionDataProvider implements ContextAwareCollectionDataProviderInterface
+{
+    private $dataProviders = [];
+    private $context = [];
+    private $providersResponse = [];
+
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider)
+    {
+        if ($collectionDataProvider instanceof ChainCollectionDataProvider) {
+            $this->dataProviders = $collectionDataProvider->dataProviders;
+        }
+    }
+
+    public function getProvidersResponse(): array
+    {
+        return $this->providersResponse;
+    }
+
+    public function getContext(): array
+    {
+        return $this->context;
+    }
+
+    public function getCollection(string $resourceClass, string $operationName = null, array $context = [])
+    {
+        $this->context = $context;
+        $results = null;
+        $match = false;
+
+        foreach ($this->dataProviders as $dataProvider) {
```

Comment:
Same here, but because it used in different classes, maybe this logic should be moved in an  `@internal` `abstract` class, or in an trait. WDYT?

Reply:
I'll need to think more about it as the logic is not really the same in the 3 providers and severeal methods doesn't use the same args number. Not sure that the result will be more readable nor maintainable. I'll check this afternoon if I can get something better. Any ideas are welcome